src/file_scanner/arrow_multi_file_info.cpp

```diff
@@ -0,0 +1,244 @@
+  auto& scan = scan_p->Cast<ArrowFileScan>();
+  auto data = make_shared_ptr<BaseUnionData>(scan_p->GetFileName());
+  if (file_idx == 0) {
+    data->names = scan.GetNames();
+    data->types = scan.GetTypes();
+    data->reader = std::move(scan_p);
+  } else {
+    data->names = scan.GetNames();
+    data->types = scan.GetTypes();
+  }
+  return data;
+}
+
+void ArrowMultiFileInfo::FinalizeReader(ClientContext& context, BaseFileReader& reader,
+                                        GlobalTableFunctionState&) {}
+
+bool ArrowMultiFileInfo::TryInitializeScan(ClientContext& context,
+                                           shared_ptr<BaseFileReader>& reader,
+                                           GlobalTableFunctionState& gstate_p,
+                                           LocalTableFunctionState& lstate_p) {
+  auto& gstate = gstate_p.Cast<ArrowFileGlobalState>();
+  auto& lstate = lstate_p.Cast<ArrowFileLocalState>();
+  if (gstate.files.find(reader->file_list_idx.GetIndex()) != gstate.files.end()) {
+    return false;
+  }
+  gstate.files.insert(reader->file_list_idx.GetIndex());
+
+  lstate.file_scan = shared_ptr_cast<BaseFileReader, ArrowFileScan>(reader);
+  lstate.local_arrow_function_data = make_uniq<ArrowScanFunctionData>(
+      &FileIPCStreamFactory::Produce,
+      reinterpret_cast<uintptr_t>(lstate.file_scan->factory.get()));
+  lstate.local_arrow_function_data->schema_root = lstate.file_scan->schema_root;
+  lstate.local_arrow_function_data->arrow_table = lstate.file_scan->arrow_table_type;
+  if (!reader->column_indexes.empty()) {
+    lstate.init_input = make_uniq<TableFunctionInitInput>(
+        *lstate.local_arrow_function_data, reader->column_indexes,
+        gstate.global_state.projection_ids, reader->filters);
+  } else {
+    lstate.init_input = make_uniq<TableFunctionInitInput>(
+        *lstate.local_arrow_function_data, gstate.global_state.column_indexes,
+        gstate.global_state.projection_ids, reader->filters);
+  }
+  lstate.local_arrow_global_state =
+      ArrowTableFunction::ArrowScanInitGlobal(context, *lstate.init_input);
+  lstate.local_arrow_local_state =
+      ArrowTableFunction::ArrowScanInitLocal(lstate.execution_context, *lstate.init_input,
+                                             lstate.local_arrow_global_state.get());
+  lstate.table_function_input = make_uniq<TableFunctionInput>(
+      lstate.local_arrow_function_data.get(), lstate.local_arrow_local_state.get(),
+      lstate.local_arrow_global_state.get());
+  return true;
+}
+
+void ArrowMultiFileInfo::Scan(ClientContext& context, BaseFileReader& reader,
+                              GlobalTableFunctionState& global_state,
+                              LocalTableFunctionState& local_state, DataChunk& chunk) {
+  auto& lstate = local_state.Cast<ArrowFileLocalState>();
+  ArrowTableFunction::ArrowScanFunction(context, *lstate.table_function_input, chunk);
+}
+
+void ArrowMultiFileInfo::FinishFile(ClientContext& context,
```

Comment:
Similarly to the comment above, this is where you have the opportunity to reset the global state when the last file that was being scanned has divided all of the possible rowgroups amongst the available local states